### ISSUE TITLE
TAG 0.8.2，in C++ 20 make error ，in C++17 make succeed，why?

### ISSUE THREAD

**Issue by wuyaozong-qd** [conan-io/conan-center-index#27215]
[ 50%] Building CXX object CMakeFiles/websocket_client.dir/src/main.cpp.o
In file included from /home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/config/core_client.hpp:55,
                 from /home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/config/asio_client.hpp:31,
                 from /home/wyz/project/websocket_client/src/main.cpp:2:
/home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/logger/basic.hpp:61:54: error: expected ‘)’ before ‘h’
   61 |     basic<concurrency,names>(channel_type_hint::value h =
      |                             ~                        ^~
      |                                                      )
/home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/logger/basic.hpp:67:42: error: expected ‘)’ before ‘*’ token
   67 |     basic<concurrency,names>(std::ostream * out)
      |                             ~            ^~
      |                                          )
/home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/logger/basic.hpp:72:35: error: expected ‘)’ before ‘c’
   72 |     basic<concurrency,names>(level c, channel_type_hint::value h =
      |                             ~     ^~
      |                                   )
/home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/logger/basic.hpp:78:35: error: expected ‘)’ before ‘c’
   78 |     basic<concurrency,names>(level c, std::ostream * out)
      |                             ~     ^~
      |                                   )
/home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/logger/basic.hpp:84:5: error: template-id not allowed for destructor
   84 |     ~basic<concurrency,names>() {}
      |     ^
/home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/logger/basic.hpp:87:54: error: expected ‘)’ before ‘const’
   87 |     basic<concurrency,names>(basic<concurrency,names> const & other)
      |                             ~                        ^~~~~~
      |                                                      )
/home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/logger/basic.hpp:100:54: error: expected ‘)’ before ‘&&’ token
  100 |     basic<concurrency,names>(basic<concurrency,names> && other)
      |                             ~                        ^~~
      |                                                      )
In file included from /home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/roles/client_endpoint.hpp:31,
                 from /home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/client.hpp:31,
                 from /home/wyz/project/websocket_client/src/main.cpp:3:
/home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/endpoint.hpp:112:5: error: template-id not allowed for destructor
  112 |     ~endpoint<connection,config>() {}
      |     ^
/home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/endpoint.hpp: In instantiation of ‘websocketpp::endpoint<connection, config>::endpoint(bool) [with connection = websocketpp::connection<websocketpp::config::asio_tls_client>; config = websocketpp::config::asio_tls_client]’:
/home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/roles/client_endpoint.hpp:72:44:   required from ‘websocketpp::client<config>::client() [with config = websocketpp::config::asio_tls_client]’
/home/wyz/project/websocket_client/src/main.cpp:19:8:   required from here
/home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/endpoint.hpp:92:16: error: no matching function for call to ‘websocketpp::log::basic<websocketpp::concurrency::basic, websocketpp::log::alevel>::basic(const level&, const value&)’
   92 |       : m_alog(new alog_type(config::alog_level, log::channel_type_hint::access))
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/config/core_client.hpp:55,
                 from /home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/config/asio_client.hpp:31,
                 from /home/wyz/project/websocket_client/src/main.cpp:2:
/home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/logger/basic.hpp:59:7: note: candidate: ‘websocketpp::log::basic<websocketpp::concurrency::basic, websocketpp::log::alevel>::basic()’ (deleted)
   59 | class basic {
      |       ^~~~~
/home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/logger/basic.hpp:59:7: note:   candidate expects 0 arguments, 2 provided
/home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/logger/basic.hpp:59:7: note: candidate: ‘websocketpp::log::basic<websocketpp::concurrency::basic, websocketpp::log::alevel>::basic(const websocketpp::log::basic<websocketpp::concurrency::basic, websocketpp::log::alevel>&)’ (deleted)
/home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/logger/basic.hpp:59:7: note:   candidate expects 1 argument, 2 provided
In file included from /home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/roles/client_endpoint.hpp:31,
                 from /home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/client.hpp:31,
                 from /home/wyz/project/websocket_client/src/main.cpp:3:
/home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/endpoint.hpp:93:16: error: no matching function for call to ‘websocketpp::log::basic<websocketpp::concurrency::basic, websocketpp::log::elevel>::basic(const level&, const value&)’
   93 |       , m_elog(new elog_type(config::elog_level, log::channel_type_hint::error))
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/config/core_client.hpp:55,
                 from /home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/config/asio_client.hpp:31,
                 from /home/wyz/project/websocket_client/src/main.cpp:2:
/home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/logger/basic.hpp:59:7: note: candidate: ‘websocketpp::log::basic<websocketpp::concurrency::basic, websocketpp::log::elevel>::basic()’ (deleted)
   59 | class basic {
      |       ^~~~~
/home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/logger/basic.hpp:59:7: note:   candidate expects 0 arguments, 2 provided
/home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/logger/basic.hpp:59:7: note: candidate: ‘websocketpp::log::basic<websocketpp::concurrency::basic, websocketpp::log::elevel>::basic(const websocketpp::log::basic<websocketpp::concurrency::basic, websocketpp::log::elevel>&)’ (deleted)
/home/wyz/project/websocket_client/build/_deps/websocketpp-src/websocketpp/logger/basic.hpp:59:7: note:   candidate expects 1 argument, 2 provided
make[2]: *** [CMakeFiles/websocket_client.dir/build.make:76：CMakeFiles/websocket_client.dir/src/main.cpp.o] 错误 1
make[1]: *** [CMakeFiles/Makefile2:133：CMakeFiles/websocket_client.dir/all] 错误 2
make: *** [Makefile:136：all] 错误 2

> [](url)

**Comment by wuyaozong-qd** [comment]
`// main.cpp
#include <websocketpp/config/asio_client.hpp>
#include <websocketpp/client.hpp>
#include <nlohmann/json.hpp>
#include <iostream>
#include <chrono>
#include <thread>
#include <atomic>
#include <mutex>

using websocketpp::lib::placeholders::_1;
using websocketpp::lib::placeholders::_2;
using websocketpp::lib::bind;
using json = nlohmann::json;

typedef websocketpp::client<websocketpp::config::asio_tls_client> client;
typedef websocketpp::lib::shared_ptr<websocketpp::lib::asio::ssl::context> context_ptr;

client c;
websocketpp::connection_hdl hdl;
std::atomic<bool> connected(false);
std::atomic<bool> timer_stop(false);
std::mutex connection_mutex;

const std::string deviceNo = "DEV_DEVICE_TEST_123";

std::string create_message() {
    json msg = {
        {"action", "device_text_stream"},
        {"bizCode", "toy"},
        {"data", "{\"audioText\":\"在吗？\",\"audioBase64\":\"\"}"},
        {"deviceNo", deviceNo},
        {"digest", "a201e085d46948a319a0b71ad66f371e"},
        {"endSign", "end"},
        {"extraMap", {
            {"sampleRate", "1600510"},
            {"encoding", "value2"},
            {"memberId", "value3"}
        }},
        {"source", "DEVICE"},
        {"timestamp", "1727077002365"}
    };
    return msg.dump();
}

context_ptr on_tls_init() {
     // 创建 SSL 上下文
    auto ctx = websocketpp::lib::make_shared<websocketpp::lib::asio::ssl::context>(
        websocketpp::lib::asio::ssl::context::sslv23
    );

    try {
        // 设置 SSL 选项（禁用 SSLv2/SSLv3，启用现代协议）
        ctx->set_options(
            websocketpp::lib::asio::ssl::context::default_workarounds |
            websocketpp::lib::asio::ssl::context::no_sslv2 |
            websocketpp::lib::asio::ssl::context::no_sslv3 |
            websocketpp::lib::asio::ssl::context::single_dh_use  // 推荐添加此选项
        );

        ctx->set_default_verify_paths();
    } catch (std::exception& e) {
        std::cerr << "TLS init error: " << e.what() << std::endl;
    }
    return ctx;
}

void message_sender() {
    std::string message = create_message();
    while (!timer_stop) {
        if (connected) {
            try {
                std::lock_guard<std::mutex> lock(connection_mutex);
                c.send(hdl, message, websocketpp::frame::opcode::text);
                std::cout << "Message sent: " << message << std::endl;
            } catch (const std::exception& e) {
                std::cerr << "Send error: " << e.what() << std::endl;
            }
        }
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }
}

void reconnect() {
    std::lock_guard<std::mutex> lock(connection_mutex);
    try {
        websocketpp::lib::error_code ec;
        std::string uri = "wss://toy-dgw-server-dev.quplusplus.net/ws/dgw-secure?deviceNo=" + deviceNo;
        client::connection_ptr con = c.get_connection(uri, ec);

        if (ec) {
            std::cerr << "Connection error: " << ec.message() << std::endl;
            return;
        }

        c.connect(con);
    } catch (const std::exception& e) {
        std::cerr << "Reconnect failed: " << e.what() << std::endl;
    }
}

int main() {
    try {
        c.clear_access_channels(websocketpp::log::alevel::all);
        c.init_asio();
        c.set_tls_init_handler(bind(&on_tls_init));

        c.set_open_handler([&](websocketpp::connection_hdl handle) {
            std::lock_guard<std::mutex> lock(connection_mutex);
            hdl = handle;
            connected = true;
            std::cout << "Connected to server" << std::endl;
        });

        c.set_message_handler([&](websocketpp::connection_hdl, client::message_ptr msg) {
            std::cout << "Received: " << msg->get_payload() << std::endl;
        });

        c.set_close_handler([&](websocketpp::connection_hdl) {
            connected = false;
            std::cout << "Connection closed. Attempting reconnect..." << std::endl;
            reconnect();
        });

        c.set_fail_handler([&](websocketpp::connection_hdl) {
            connected = false;
            std::cerr << "Connection failed. Attempting reconnect..." << std::endl;
            reconnect();
        });

        reconnect();

        std::thread sender_thread(message_sender);
        sender_thread.detach();

        c.run();
    } catch (const std::exception& e) {
        std::cerr << "Fatal error: " << e.what() << std::endl;
        return 1;
    }
    return 0;
}`

**Comment by wuyaozong-qd** [comment]
`# CMakeLists.txt
cmake_minimum_required(VERSION 3.24)
project(websocket_client)

set(CMAKE_CXX_STANDARD 20)

find_package(OpenSSL REQUIRED)
find_package(Boost 1.74.0 REQUIRED COMPONENTS system)

include(FetchContent)
FetchContent_Declare(
        websocketpp
        GIT_REPOSITORY https://github.com/zaphoyd/websocketpp.git
        GIT_TAG 0.8.2
)
FetchContent_Declare(json URL https://github.com/nlohmann/json/releases/download/v3.11.3/json.tar.xz)

# 获取依赖项
FetchContent_MakeAvailable(websocketpp json)


include_directories(
        ${OPENSSL_INCLUDE_DIR}
        # Path to websocketpp and json headers
        ${WEBSOCKETPP_SOURCE_DIR}
        /usr/local/include/
        ${CMAKE_CURRENT_SOURCE_DIR}/include/
)

add_executable(websocket_client src/main.cpp)

# 包含目录配置（关键修正）
target_include_directories(${PROJECT_NAME} PRIVATE
        ${websocketpp_SOURCE_DIR}  # 正确包含WebSocket++头文件路径
)
target_link_libraries(websocket_client
        PRIVATE
        OpenSSL::SSL
        OpenSSL::Crypto
        nlohmann_json::nlohmann_json
)`

**Comment by Nekto89** [comment]
how is this related to conan-center-index?

**Comment by jcar87** [comment]
Closing as can't see any evidence of Conan being used in the logs - if it's a problem with a library, I'd advise to contact the maintainers directly. If there is an issue with a _recipe_ in Conan Center, let us know.
